File: 133_CountDistinctSubstrings.cpp
/*
Problem: Count Distinct Substrings
Description: Count number of distinct substrings of a string.

Approach: Insert all suffixes into trie. Each node represents a unique substring.
          Count total nodes in trie.
*/

#include <bits/stdc++.h>
using namespace std;

class TrieNode {
public:
    TrieNode* children[26];
    
    TrieNode() {
        for (int i = 0; i < 26; i++)
            children[i] = nullptr;
    }
};

int countDistinctSubstrings(string s) {
    TrieNode* root = new TrieNode();
    int count = 0;  // Empty string
    
    // Insert all suffixes
    for (int i = 0; i < s.size(); i++) {
        TrieNode* node = root;
        
        for (int j = i; j < s.size(); j++) {
            int idx = s[j] - 'a';
            
            if (!node->children[idx]) {
                node->children[idx] = new TrieNode();
                count++;  // New unique substring
            }
            
            node = node->children[idx];
        }
    }
    
    return count + 1;  // +1 for empty string
}

// Alternative: Using set
int countDistinctSubstringsSet(string s) {
    unordered_set<string> uniqueSubstrings;
    
    for (int i = 0; i < s.size(); i++) {
        for (int j = i; j < s.size(); j++) {
            uniqueSubstrings.insert(s.substr(i, j - i + 1));
        }
    }
    
    return uniqueSubstrings.size() + 1;  // +1 for empty string
}

int main() {
    string s1 = "abab";
    cout << "Distinct substrings (Trie): " << countDistinctSubstrings(s1) << endl;  // Output: 8
    cout << "Distinct substrings (Set): " << countDistinctSubstringsSet(s1) << endl;
    
    string s2 = "abc";
    cout << "Distinct substrings: " << countDistinctSubstrings(s2) << endl;  // Output: 7
    
    return 0;
}

/*
Time Complexity: O(n^2)
Space Complexity: O(n^2)
*/

File: 133_CountDistinctSubstrings.py
"""
Problem: Count Distinct Substrings
Description: Count number of distinct substrings of a string.

Approach: Insert all suffixes into trie. Each node represents a unique substring.
          Count total nodes in trie.
"""

class TrieNode:
    def __init__(self):
        self.children = {}

def countDistinctSubstrings(s: str) -> int:
    root = TrieNode()
    count = 0  # Empty string
    
    # Insert all suffixes
    for i in range(len(s)):
        node = root
        
        for j in range(i, len(s)):
            if s[j] not in node.children:
                node.children[s[j]] = TrieNode()
                count += 1  # New unique substring
            
            node = node.children[s[j]]
    
    return count + 1  # +1 for empty string

# Alternative: Using set
def countDistinctSubstringsSet(s: str) -> int:
    uniqueSubstrings = set()
    
    for i in range(len(s)):
        for j in range(i, len(s)):
            uniqueSubstrings.add(s[i:j+1])
    
    return len(uniqueSubstrings) + 1  # +1 for empty string

def main():
    s1 = "abab"
    print("Distinct substrings (Trie):", countDistinctSubstrings(s1))  # Output: 8
    print("Distinct substrings (Set):", countDistinctSubstringsSet(s1))
    
    s2 = "abc"
    print("Distinct substrings:", countDistinctSubstrings(s2))  # Output: 7

if __name__ == "__main__":
    main()

"""
Time Complexity: O(n^2)
Space Complexity: O(n^2)
"""
